libraries/AP_Math/tests/test_math.cpp

```diff
@@ -0,0 +1,309 @@
+{  
+  for(int i = 0; i < 1000; i++) {
+      if(i < 250) {
+          EXPECT_EQ(constrain_float(i, 250, 500), 250);
+          EXPECT_EQ(constrain_int16(i, 250, 500), 250);
+          EXPECT_EQ(constrain_int32(i, 250, 500), 250);
+      }
+      else if(i > 500) {
+          EXPECT_EQ(constrain_float(i, 250, 500), 500);
+          EXPECT_EQ(constrain_int16(i, 250, 500), 500);
+          EXPECT_EQ(constrain_int32(i, 250, 500), 500);
+      }
+      else {
+          EXPECT_EQ(constrain_float(i, 250, 500), i);        
+          EXPECT_EQ(constrain_int16(i, 250, 500), i);
+          EXPECT_EQ(constrain_int32(i, 250, 500), i);
+      }
+  }
+  
+  for(int i = 0; i <= 1000; i++) {
+      int c = i - 1000;
+      if(c < -250) {
+          EXPECT_EQ(constrain_float(c, -250, -50), -250);
+          EXPECT_EQ(constrain_int16(c, -250, -50), -250);
+          EXPECT_EQ(constrain_int32(c, -250, -50), -250);
+      }
+      else if(c > -50) {
+          EXPECT_EQ(constrain_float(c, -250, -50), -50);
+          EXPECT_EQ(constrain_int16(c, -250, -50), -50);
+          EXPECT_EQ(constrain_int32(c, -250, -50), -50);
+      }
+      else {
+          EXPECT_EQ(constrain_float(c, -250, -50), c);        
+          EXPECT_EQ(constrain_int16(c, -250, -50), c);
+          EXPECT_EQ(constrain_int32(c, -250, -50), c);
+      }
+  }
+  
+  for(int i = 0; i <= 2000; i++) {
+      int c = i - 1000;
+      if(c < -250) {
+          EXPECT_EQ(constrain_float(c, -250, 50), -250);
+          EXPECT_EQ(constrain_int16(c, -250, 50), -250);
+          EXPECT_EQ(constrain_int32(c, -250, 50), -250);
+      }
+      else if(c > 50) {
+          EXPECT_EQ(constrain_float(c, -250, 50), 50);
+          EXPECT_EQ(constrain_int16(c, -250, 50), 50);
+          EXPECT_EQ(constrain_int32(c, -250, 50), 50);
+      }
+      else {
+          EXPECT_EQ(constrain_float(c, -250, 50), c);        
+          EXPECT_EQ(constrain_int16(c, -250, 50), c);
+          EXPECT_EQ(constrain_int32(c, -250, 50), c);
+      }
+  }
+}
+
+TEST(Wrap180Test, angle180)
+{
+  // Full circle test
+  for(int i = 0; i < 360; i++) {
+      // smaller pole position
+      if(i < 180) {
+          EXPECT_EQ(wrap_180(i), i);
+          EXPECT_EQ(wrap_180(-i), -i);
+      }
+      // hit pole position -180/+180 degree
+      else if(i == 180) {
+          EXPECT_EQ(wrap_180(i), i);
+          EXPECT_EQ(wrap_180(-i), i);
+      }
+      // bigger pole position
+      else {
+          EXPECT_EQ(wrap_180(i), -(360-i));
+          EXPECT_EQ(wrap_180(-i), (360-i));
+      }
+  }
+  
+  EXPECT_EQ(wrap_180(45.f),        45.f);
+  EXPECT_EQ(wrap_180(90.f),        90.f);
+  EXPECT_EQ(wrap_180(180.f),       180.f);
+  EXPECT_EQ(wrap_180(180.1f),     -179.9f);
+  EXPECT_EQ(wrap_180(270.f),      -90.f);
+  EXPECT_EQ(wrap_180(360.f),       0.f);
+  EXPECT_EQ(wrap_180(720.f),       0.f);
+  EXPECT_EQ(wrap_180(3600.f),      0.f);
+  EXPECT_EQ(wrap_180(7200.f),      0.f);
+  EXPECT_EQ(wrap_180(-36000000.f), 0.f);
```

Comment:
@lucasdemarchi 
I'm getting this error message, which presumably means I have to build gtest for px4?
```
andy@andy-netbook:~/cpp/projects/apmwaf/ardupilot$ export PATH=$PATH:/opt/gcc-arm-none-eabi-4_7-2013q2/bin
andy@andy-netbook:~/cpp/projects/apmwaf/ardupilot$ waf configure --board=px4-v2
Setting top to                           : /home/andy/cpp/projects/apmwaf/ardupilot 
Setting out to                           : /home/andy/cpp/projects/apmwaf/ardupilot/build 
Setting board to                         : px4-v2 
Using toolchain                          : arm-none-eabi 
Checking for 'g++' (C++ compiler)        : arm-none-eabi-g++ 
Checking for 'gcc' (C compiler)          : arm-none-eabi-gcc 
Checking for HAVE_CMATH_ISFINITE         : no 
Checking for HAVE_CMATH_ISINF            : no 
Checking for HAVE_CMATH_ISNAN            : no 
Checking for NEED_CMATH_ISFINITE_STD_NAMESPACE : no 
Checking for NEED_CMATH_ISINF_STD_NAMESPACE    : no 
Checking for NEED_CMATH_ISNAN_STD_NAMESPACE    : no 
Checking for program 'cmake'                   : /usr/bin/cmake 
Checking for program 'ninja, ninja-build'      : /usr/local/bin/ninja 
Checking for program 'cp'                      : /bin/cp 
Checking for program 'python'                  : /usr/bin/python 
Checking for python version                    : (2, 7, 3, 'final', 0) 
Checking for program 'git'                     : /usr/bin/git 
Gtest                                          : PX4 boards currently don't support compiling gtest 
Benchmarks                                     : disabled 
Unit tests                                     : disabled 
'configure' finished successfully (7.206s)
andy@andy-netbook:~/cpp/projects/apmwaf/ardupilot$ waf check
Waf: Entering directory `/home/andy/cpp/projects/apmwaf/ardupilot/build/px4-v2'
check: gtest library is required
```